fms/models/llava_next.py

```diff
@@ -0,0 +1,586 @@
+
+        if config is not None:
+            self.config = config
+        else:
+            self.config = LlavaNextConfig()
+        self.config = self.config.updated(**kwargs)
+        self.distributed_strategy = distributed_strategy
+
+        if not isinstance(self.config.vision_config, SiglipVisionConfig):
+            print(
+                "FMS implementation of LlavaNext currently supports only Siglip vision model"
+            )
+        if not isinstance(self.config.text_config, GraniteConfig):
+            print(
+                "FMS implementation of LlavaNext currently supports only Granite language model"
+            )
+
+        # Only supporting granite text decoder encoder for now
+        self.language_model = Granite(self.config.text_config)
+
+        # Only supporting siglip vision encoder for now
+        self.vision_tower = SiglipVision(self.config.vision_config)
+
+        self.multi_modal_projector = LlavaNextMultiModalProjector(self.config)
+        embed_std = 1 / math.sqrt(self.config.text_config.emb_dim)
+        self.image_newline = nn.Parameter(
+            torch.randn(self.config.text_config.emb_dim) * embed_std
+        )
+        self.vocab_size = self.config.text_config.src_vocab_size
+
+    @classmethod
+    def from_config(cls, config: LlavaNextConfig) -> "LlavaNext":
+        return cls(config)
+
+    def get_config(self) -> LlavaNextConfig:
+        return self.config
+
+    def reset_parameters(self):
+        nn.init.xavier_uniform_(self.image_newline.data)
+        self.langauage_model.reset_parameters()
+        self.vision_tower.reset_parameters()
+        self.multi_modal_projector.reset_parameters()
+
+    def post_init(self):
+        self.language_model.post_init()
+        self.vision_tower.post_init()
+
+    def unpad_image(self, tensor, original_size):
+        if not isinstance(original_size, (list, tuple)):
+            original_size = original_size.tolist()
+        original_height, original_width = original_size
+        current_height, current_width = tensor.shape[1:]
+
+        original_aspect_ratio = original_width / original_height
+        current_aspect_ratio = current_width / current_height
+
+        if original_aspect_ratio > current_aspect_ratio:
+            scale_factor = current_width / original_width
+            new_height = int(round(original_height * scale_factor, 7))
+            padding = (current_height - new_height) // 2
+            unpadded_tensor = tensor[:, padding : current_height - padding, :]
+        else:
+            scale_factor = current_height / original_height
+            new_width = int(round(original_width * scale_factor, 7))
+            padding = (current_width - new_width) // 2
+            unpadded_tensor = tensor[:, :, padding : current_width - padding]
+
+        return unpadded_tensor
+
+    # TODO: fix graph break in the HF impl here
+    def select_best_resolution(self, original_size, possible_resolutions: list):
```

Comment:
I think if we're careful with tensor arithmetic, all of this can be rewritten to not have graph breaks. Let's set up a short discussion offline to figure it out.

Reply:
in case we can't figure it out, and given you already have a custom function hook for generate(), the call to image_size_to_num_patches can be moved there and outside of the forward() pass, which will remove the graph break, and then we change the inputs to forward from image_sizes toimage_num_patches

Reply:
Yup, I believe so too. First pass was HF equivalence, followed by graph break removal, followed by alternative ops for AIU.

Good idea regarding the custom function hook fallback, will keep in mind!